### ISSUE TITLE
Elasticsearch: Revert terms min_doc_count default from 0 back to 1

### ISSUE THREAD

**Issue by csoulios** [grafana/grafana#29616]
**What would you like to be added**:

When running `terms` aggregation in elasticsearch, the default value of `min_doc_count` has been set to `0` (#24204). This changed the default behavior of `terms` aggregations and returns empty buckets (terms with `doc_count: 0`)

**Why is this needed**:

Issue #24204 reported: 

> When graphing queries from an elastic datasource I am unable to set the min_doc_count to 0 on terms query. The interface always resets this value to 1 when trying to select 0. This allows us to view the terms that have 0 value. 

In an effort to fix issue #23636, PR #24204 set the default value of `min_doc_count` to `0`. 

However, I think what the issue actually reported was that users were not able to override the default `min_doc_count` setting which back then was `1`.  

I would suggest that default `min_doc_count: 1` is preferred, as current implementation returns buckets that have no data. There are use cases that this returns a lot of irrelevant buckets.

**Comment by dagwieers** [comment]
I agree, this behavior confused us a lot. To me the default behavior was either a bug or an oversight that required filter options to the terms lookup. Once I discovered that the min_doc_count actually covered most (not all) of our use-cases I also wondered why this is not the default behavior.

Thanks for opening this ticket!

**Comment by Elfo404** [comment]
as per https://github.com/grafana/grafana/issues/6121#issuecomment-249279961, i think it actually makes sense to have `min_doc_count: 0` as a default, although I can see why this might be confusing. 

I'm closing this issue for now, also because changing again the default behavior might be even more confusing and this didn't receive much feedback, but we can for sure discuss it further if you think it's needed.

Thanks for your contributions!

**Comment by Ultraseamus** [comment]
I wrote a bug for this, where I explain why the default makes much more sense as 1 than 0, that was closed in favor of this one.

https://github.com/grafana/grafana/issues/36865

@Elfo404 The comment you linked was from 2016. And since then, the default had been switched to 1 to match up with Elasticsearch, which was the right choice (the bug was that it was impossible to switch back to 0, and IMO I thought that was on purpose since it almost never makes sense for a terms aggregation to have a min doc count of 0).

This value should be allowed to be set to 0, but the default should match Elasticsearch's, they defaulted it to 1 for a reason. The Elasticsearch documentation even includes a warning section explaining why the default is 1 and not 0.


> Setting min_doc_count=0 will also return buckets for terms that didn’t match any hit. However, some of the returned terms which have a document count of zero might only belong to deleted documents or documents from other types, so there is no warranty that a match_all query would find a positive document count for those terms.

A setting of 0 is such a rare, niche case. If you filtered your query down to 1 term, but were using 0 as your min, you would get a result back with more than 1 term. It returns results that specifically do not match your filter. And on top of that, a 0 doc count here means you could be getting back results from deleted documents.

I am very confident that the default here should be 1. If the examples from my other comments and here are not convincing enough, I think it is clear that the policy should be to match what the Elastic team sets as their default. I'm pushing for it because the default value of 0 is confusing for users. Terms they filtered out will show up, even delated data in the index can show up. '0' is an advanced edge case that 99% of users will never want.

Just to be clear, this is different from the Date Histogram min doc count, which is and should be defaulted to 0.

**Comment by Elfo404** [comment]
@Ultraseamus thanks for taking the time to follow up on this, i probably overlooked this issue.

I'm reopening this and be sure to take some time to look at it.

**Comment by Ultraseamus** [comment]
Thank you for reopening it.

I rambled a bit in my comment above, but I think the cleanest argument for changing it back is simple: If a user does not change an optional setting on an aggregation, it should be treated as if that option was null (meaning ES will run with its defaults). There needs to be a very good reason for Grafana to select a non-default option for a user.
